Add dynamic loading support

joechrisellis opened the pull request:
Hi guys,

Following on from [issue 1541](https://github.com/rust-lang/rust-bindgen/issues/1541), here's an implementation for libloading support in bindgen. I'd still consider this a prototype at the minute, so any comments are welcome and appreciated.

# Implementation

- The existing functionality is maintained if dynamic loading mode is not specified.
- We introduce two new flags -- `--dynamic-loading` and `--dynamic-library-name` -- which enable dynamic loading mode and specify a name for the shared library respectively. The 'name' of the shared library is what you want the struct containing the bindings to be called. For example, you could call it `BZip2` -- then you'd instantiate the dynamically loaded library with `let lib = BZip2::new();`. If `--dynamic-library-name` is not specified it'll just default to `Lib`.
- If dynamic loading mode is specified, the bindings are created inside a library struct alongside some libloading boilerplate.

# Example

Given the following `header.h`:

```
int foo(int x);
void bar(double x);
```

And the following code inside our `build.rs`:

```
let bindings = bindgen::Builder::default()
    .header("header.h")
    // Use dynamic loading
    .dynamic_loading(true)
    .dynamic_library_name("MyLibrary")
    // ...
    .generate()
    .expect("Unable to generate bindings");
```

We generate the following bindings:

```
/* automatically generated by rust-bindgen 0.54.1 */

pub struct MyLibrary<'a> {
    foo: libloading::Symbol<
        'a,
        unsafe extern "C" fn(x: ::std::os::raw::c_int) -> ::std::os::raw::c_int,
    >,
    bar: libloading::Symbol<'a, unsafe extern "C" fn(x: f64)>,
}
impl<'a> MyLibrary<'a> {
    pub fn new(lib: &libloading::Library) -> MyLibrary {
        unsafe {
            MyLibrary {
                foo: lib.get("foo".as_bytes()).unwrap(),
                bar: lib.get("bar".as_bytes()).unwrap(),
            }
        }
    }
}
```

Which we can then use in our code with the following:

```
#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]
include!(concat!(env!("OUT_DIR"), "/bindings.rs"));

extern crate libloading;

pub fn main() {
    let lib = libloading::Library::new("/path/to/shared/lib.so").unwrap();
    let library_wrapper = MyLibrary::new(&lib);
    unsafe { (library_wrapper.bar)(123.0) };
}
```

## Multiple Dynamically Loaded Libraries

If you want to use multiple dynamically loaded libraries, the correct approach would be to do this in your `build.rs`:

```
let bindings_x = bindgen::Builder::default()
    .header("header_x.h")
    // Use dynamic loading
    .dynamic_loading(true)
    .dynamic_library_name("LibraryX")
    // ...
    .generate()
    .expect("Unable to generate bindings");

let bindings_y = bindgen::Builder::default()
    .header("header_y.h")
    // Use dynamic loading
    .dynamic_loading(true)
    .dynamic_library_name("LibraryY")
    // ...
    .generate()
    .expect("Unable to generate bindings");

let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
lib1_bindings
    .write_to_file(out_path.join("libx_bindings.rs"))
    .expect("Couldn't write bindings!");
lib2_bindings
    .write_to_file(out_path.join("liby_bindings.rs"))
    .expect("Couldn't write bindings!");
```

Then do the following in your code:

```
extern crate libloading;

mod libx {
    include!(concat!(env!("OUT_DIR"), "/libx_bindings.rs"));
}
mod liby {
    include!(concat!(env!("OUT_DIR"), "/liby_bindings.rs"));
}
pub use lib1::LibraryX;
pub use lib2::LibraryY;
```

(the `mod libx { ... }` is to prevent namespace collisions if `LibraryX` and `LibraryY` have common includes -- e.g. if they both have `#include <stdint.h>`)

---

Comments and criticisms absolutely welcome! If this seems like a suitable approach I can update the docs in a separate PR. Cheers! 🍻

dergroncki commented:
I tested the dynamic loading support with one dll file on windows and it worked like a charm! Exactly what I wanted.

Question: Instead of "MyLibrary" as in your example I used "ModelLib", but "ModelLib" is marked as unknown although the program is compiled without errors. Thanks for any hint.

![Screenshot 2020-08-02 at 17 32 48](https://user-images.githubusercontent.com/2815398/89126597-16023a00-d4e7-11ea-9faf-783586498c65.jpg)

Program extract of the automatically generated code:

```rust
extern crate libloading;
pub struct ModelLib<'a> {
    #[doc = ""]
    createInstance: libloading::Symbol<
        'a,
        unsafe extern "C" fn(lengthBillet_m: f64, radiusBillet_m: f64) -> *mut ModelData,
    >,
    destroyInstance:
        libloading::Symbol<'a, unsafe extern "C" fn(data: *mut ModelData) -> ::std::os::raw::c_int>,
    calculate:
        libloading::Symbol<'a, unsafe extern "C" fn(data: *mut ModelData) -> ::std::os::raw::c_int>,
    enableLogging: libloading::Symbol<
        'a,
        unsafe extern "C" fn(
            data: *mut ModelData,
            path: *const ::std::os::raw::c_char,
        ) -> ::std::os::raw::c_int,
    >,
    errorLog: libloading::Symbol<
        'a,
        unsafe extern "C" fn(msg: *mut ::std::os::raw::c_char, maxSize: ::std::os::raw::c_int),
    >,
    errorLogSize: libloading::Symbol<'a, unsafe extern "C" fn() -> ::std::os::raw::c_int>,
    customer: libloading::Symbol<
        'a,
        unsafe extern "C" fn(msg: *mut ::std::os::raw::c_char, maxSize: ::std::os::raw::c_int),
    >,
    date: libloading::Symbol<
        'a,
        unsafe extern "C" fn(msg: *mut ::std::os::raw::c_char, maxSize: ::std::os::raw::c_int),
    >,
    version: libloading::Symbol<
        'a,
        unsafe extern "C" fn(msg: *mut ::std::os::raw::c_char, maxSize: ::std::os::raw::c_int),
    >,
}
impl<'a> ModelLib<'a> {
    pub fn new(lib: &libloading::Library) -> ModelLib {
        unsafe {
            ModelLib {
                createInstance: lib.get("createInstance".as_bytes()).unwrap(),
                destroyInstance: lib.get("destroyInstance".as_bytes()).unwrap(),
                calculate: lib.get("calculate".as_bytes()).unwrap(),
                enableLogging: lib.get("enableLogging".as_bytes()).unwrap(),
                errorLog: lib.get("errorLog".as_bytes()).unwrap(),
                errorLogSize: lib.get("errorLogSize".as_bytes()).unwrap(),
                customer: lib.get("customer".as_bytes()).unwrap(),
                date: lib.get("date".as_bytes()).unwrap(),
                version: lib.get("version".as_bytes()).unwrap(),
            }
        }
    }
}
```

joechrisellis commented:
@dergroncki, good to know that it worked for you! If the program compiles just fine, then the 'not found' error you're seeing is probably because your IDE is unable to process the `include` macros for whatever reason. You could test this by copying the generated `bindings.rs` file and dumping it into your file in the place of the `include` macro -- if this doesn't show any errors, then it's probably that.

On another, rather more general note -- I'm planning on spending some time today to get this PR into a production-ready state. At the moment, it's all rather ad-hoc and will occasionally generate invalid bindings for C++ headers. Happy to see that you have found it useful though. 😄

joechrisellis commented on src/ir/context.rs:
```diff
@@ -2412,6 +2428,70 @@ If you encounter an error missing from this list, please file an issue or a PR!"
+                // - The item being considered is at the toplevel (but is not the root itself).
+                // - The item is permitted by the {white,black}list.
+
+                // Finally, only generate dynamic bindings for functions (for now! this could
+                // change in the future to support variables).
```
Yeah, AFAIK static constants should be dynamically-loadable too, but I haven't thought about this too much. 👍

joechrisellis commented on src/lib.rs:
```diff
@@ -1699,6 +1727,12 @@ struct BindgenOptions {
     wasm_import_module_name: Option<String>,
+
+    /// Whether or not we are generating bindings for a shared library.
+    dynamic_loading: bool,
+
+    /// The name of the dynamic library if we are generating bindings for a shared library.
+    dynamic_library_name: Option<String>,
```
Yes, that's a much better idea. I'll do this 😄

goto-bus-stop commented:
This looks nice! I'm trying to use it with a two-crate setup, one -sys and one Rusty wrapper. Unfortunately that doesn't work right now because the struct members are private. Should they just all be declared `pub` by default? I think that would match the non-dynamic bindgen behaviour.

joechrisellis commented:
> This looks nice! I'm trying to use it with a two-crate setup, one -sys and one Rusty wrapper. Unfortunately that doesn't work right now because the struct members are private. Should they just all be declared `pub` by default? I think that would match the non-dynamic bindgen behaviour.

Yes, you're absolutely right, they should be. This is a trivial fix, will update in just a moment. My bad! 🙂

joechrisellis commented:
@nagisa thanks for your excellent suggestions! I will update the patch accordingly. 😄

joechrisellis commented:
@Michael-F-Bryan would it be okay to adapt some of your ideas for this patch? Specifically, we're interested in doing what @nagisa has suggested above, which is very similar to your work. Just wanted to check with you! 😄

timfish commented:
Many thanks for all the work that's going into getting dynamic support. I've already had to write loads of dynamic bindings manually and it's painful .

@joechrisellis if you're passing the library name to `.dynamic_library_name("LibraryY")` in `build.rs`, could you make the `path` in `load_from_path` an `Option` and default to that name is no path is supplied? This would allow users to load from a specific path or pass `None` to use the OS's default resolution. 

So modifying the work from @Michael-F-Bryan : 
```rust
impl Bindings {
    pub unsafe fn load_from_path<P>(
        path: Option<P>,
    ) -> Result<Self, ::libloading::Error>
    where
        P: AsRef<::std::ffi::OsStr>,
    {
        let path = path.unwrap_or("LibraryY"); // something like this...
        let library = ::libloading::Library::new(path)?;
        ...
    }
}
``` 

My dream would be that when I'm creating a `*-sys` crate, consumers of the crate could switch between dynamic and static with only a feature flag. This would require that both are exposed though the same code paths.

So for dynamic you'd get:
```rust
struct SomeName<'a> {
    sym1: Symbol<...>,
    sym2: Symbol<...>,
}

impl SomeName<'a> {
    unsafe extern "C" fn sym1(&self, ...) { /* call self.sym1 */ }
    unsafe extern "C" fn sym2(&self, ...) { /* call self.sym2 */ }
}
```
and for static you'd get:
```rust
struct SomeName;

impl SomeName<'a> {
    unsafe extern "C" fn sym1(&self, ...) { /* call static bindings */ }
    unsafe extern "C" fn sym2(&self, ...) { /* call static bindings */ }
}
```

Without this, the Rust wrappers for sys crates are going to need some laborious boilerplate abstraction over the top to support both dynamic and static bindings. 

I'm not suggesting we modify the default bindgen behaviour, but it would be nice to support this kind of usage.

joechrisellis commented:
Hi @timfish!

> @joechrisellis if you're passing the library name to .dynamic_library_name("LibraryY") in build.rs, could you make the path in load_from_path an Option and default to that name is no path is supplied? This would allow users to load from a specific path or pass None to use the OS's default resolution.

I am a little skeptical of overloading the `.dynamic_library_name` method -- at the moment, that sets the name of the _struct_ for the generated bindings. The purpose of this is so that we can avoid name collisions if we want to include multiple dynamically-loaded libraries in a single project. This is explained a bit more in the original issue (issue 1541), but I think some of the information there might be slightly outdated now. Unless there is some mechanism/env var like `LD_LIBRARY_PATH` that can resolve the name of the library (i.e. `LibraryY`) to a path at runtime that I am unaware of? Let me know if there is! 😄 

However, I do like the idea of perhaps passing a path at build-time in `build.rs`. Something like `.dynamic_library_path` would work quite well. So, we could do exactly what you suggest, but introduce a new option for it instead of overloading an existing one. Let me know if you think this is the way to go.

> My dream would be that when I'm creating a *-sys crate, consumers of the crate could switch between dynamic and static with only a feature flag. This would require that both are exposed though the same code paths.

This is a great idea, although I do think it's probably out-of-scope for this ticket. There might also be some further challenges with doing this. I think that'd work fine for C bindings, but for C++ bindings where we generate structs for classes, we would probably have to find some workaround.

Appreciate your comments, too! Glad to see that people are finding this useful.

Michael-F-Bryan commented:
> @Michael-F-Bryan would it be okay to adapt some of your ideas for this patch? Specifically, we're interested in doing what @nagisa has suggested above, which is very similar to your work. Just wanted to check with you! smile

Yeah, for sure! The best thing that could happen is someone reuses ideas/code from my prototype and integrates them upstream :slightly_smiling_face:

hug-dev commented:
One thing I thought about: some dynamic libraries do not implement all of the functions of their corresponding API (I saw this for the PKCS 11 Crypto API for example).
That means that a call to `new` would error out if one of the function is not implemented.
If you think this is a problem as well, I can propose the following solution: the `Library` struct could contain `Option` of function pointers instead of them directly and the `new` method would set the corresponding one to `None` if the `get` call errors out.
What do you think?

joechrisellis commented:
@hug-dev that’s an excellent point. I suppose this is _partially_ handled by the {allow,block}list functionality already in bindgen, but that’s definitely a very good idea. Will have a look at doing this. 😄

Michael-F-Bryan commented:
> If you think this is a problem as well, I can propose the following solution: the Library struct could contain Option of function pointers instead of them directly and the new method would set the corresponding one to None if the get call errors out.

I can think of some other valid use cases for this... Imagine your application uses dynamic loading as a plugin mechanism and the plugin author only wants to override certain hooks. They could expose symbols for just the hooks they care about and the system will do nothing if the other symbols aren't present.

You could implement it using a similar mechanism to the whitelist, accepting a list of regular expressions for symbols that are optional. 

Returning a `Result<T, MissingSymbol>` could work in these cases, although considering a lot of C functions use return codes, you'd end up with `Result<c_int, MissingSymbol>`, which is just a `Result<Result<(), NativeError>, MissingSymbol>` in disguise... That could have an impact on ergonomics, depending on how it's handled.

joechrisellis commented:
@hug-dev, I've made a change which hopefully addresses your comments, following some of @Michael-F-Bryan's suggestions.

The approach that I took was: instead of keeping function pointers in the library struct, we keep `Result<T, ::libloading::Error>`s instead (where `T` represents the function pointer type). On instantiation of the struct, we try to grab the symbols as usual. The difference from the earlier implementation is that any errors raised through libloading (e.g. if any symbols don't exist) should not cause the instantiation to fail. We instead keep track of _which_ symbols failed to load.

The upshot of this is that if a header file defines two functions, say `foo` and `bar`, but only `foo` is implemented in the dynlib, `foo` will be set to `Ok(T)`, whereas `bar` will be set to `Err(...)`.

The helper functions inside the library have been tweaked to be consistent with this. They now return `Result<T, ::libloading::Error>` -- i.e., `Ok` with the corresponding result if the symbol was found, and `Err` if it couldn't be found when the library was instantiated.

This should be enough to support the use of partially-defined dynamic libraries, although the onus is on the caller to make sure that the function they're calling has actually been found.

So, for the following example header:

```
int foo(int x, int y);                                                                                                                                                                                                                                                                                                                                                                                                                   
int bar(void *x);
int baz();
```

We get the following bindings generated:

```
#![allow(                                                                                                                                                                                                                                                                                                                                                                                                                                
    dead_code,
    non_snake_case,
    non_camel_case_types,
    non_upper_case_globals
)]

extern crate libloading;
pub struct TestLib {
    __library: ::libloading::Library,
    foo: Result<
        unsafe extern "C" fn(
            x: ::std::os::raw::c_int,
            y: ::std::os::raw::c_int,
        ) -> ::std::os::raw::c_int,
        ::libloading::Error,
    >,
    bar: Result<
        unsafe extern "C" fn(
            x: *mut ::std::os::raw::c_void,
        ) -> ::std::os::raw::c_int,
        ::libloading::Error,
    >,
    baz: Result<
        unsafe extern "C" fn() -> ::std::os::raw::c_int,
        ::libloading::Error,
    >,
}
impl TestLib {
    pub unsafe fn new<P>(path: P) -> Result<Self, ::libloading::Error>
    where
        P: AsRef<::std::ffi::OsStr>,
    {
        let __library = ::libloading::Library::new(path)?;
        let foo = __library.get("foo".as_bytes()).map(|sym| *sym);
        let bar = __library.get("bar".as_bytes()).map(|sym| *sym);
        let baz = __library.get("baz".as_bytes()).map(|sym| *sym);
        Ok(TestLib {
            __library: __library,
            foo,
            bar,
            baz,
        })
    }
    pub unsafe extern "C" fn foo(
        &self,
        x: ::std::os::raw::c_int,
        y: ::std::os::raw::c_int,
    ) -> Result<::std::os::raw::c_int, &::libloading::Error> {
        let sym = self.foo.as_ref()?;
        Ok((sym)(x, y))
    }
    pub unsafe extern "C" fn bar(
        &self,
        x: *mut ::std::os::raw::c_void,
    ) -> Result<::std::os::raw::c_int, &::libloading::Error> {
        let sym = self.bar.as_ref()?;
        Ok((sym)(x))
    }
    pub unsafe extern "C" fn baz(
        &self,
    ) -> Result<::std::os::raw::c_int, &::libloading::Error> {
        let sym = self.baz.as_ref()?;
        Ok((sym)())
    }
}
```

Would appreciate some eyes on this, and some comments -- I'm sure there might be a more refined approach to this.

Cheers!

hug-dev commented:
This new approach looks good to me!

timfish commented:
I've been trying this out today and it works well. 

I am getting a load of warnings along the lines of:

![image](https://user-images.githubusercontent.com/1150298/92951384-2ac2dd80-f456-11ea-90ac-fa6fe46baa2d.png)

Would you emit `#![allow(improper_ctypes_definitions)]` into the top of the generated bindings or are warnings left for consumers to suppress in their `lib.rs`?

Michael-F-Bryan commented:
The generated bindings will only ever be called from Rust so they shouldn't need to be `extern "C"`.

Removing the `extern "C"` would be better than silencing a warning telling us that, quite legitimately, that `Result` isn't safe for use with FFI (Rust enums have no specified binary representation).

joechrisellis commented:
@Michael-F-Bryan @timfish -- excellent spot, this should be fixed now.

goto-bus-stop commented:
I am not sure if returning a `Result<>` is the correct approach for symbols that may not exist. That is a case that you might want to check for ahead of time, especially if the function you're calling is expected to have side effects. Less importantly, it's also a bit unfortunate to pay the cost of that check on every function call, even if you know for sure that the symbol _does_ exist.

I don't know what a good solution would be though—adding a method to check for symbol existence might be too much auto-generated API surface for bindgen, and you'd have to deal with the possibility of loaded symbol names clashing with the bindgen-provided method names…
Another option would be to go back to the public `Result<>` fields approach, so folks can do `if api_struct.my_optional_hook.is_ok() {}`, but then it's back to being strongly tied to the libloading API :/

joechrisellis commented:
@goto-bus-stop -- I do agree, it is a bit of a pain to have to pay the cost of checking a `Result<>` on every function call. Similarly, though, I am unaware of a better approach. I think the easiest approach from a user-experience perspective would be to revert back to the way things were before, as you said. For me, at least, it seems like supporting partially-defined dynamic libs is perhaps too much of an edge case to sacrifice the usability. The user could still _use_ partially-defined dynamic libs with this patch, but they would have to make use of the {allow,block}list functionality to prune out anything that's undefined.

_Or_ perhaps we could add a switch to alternate between the two output forms. But I think this will very quickly start to get convoluted.

@hug-dev do you have any thoughts here?

hug-dev commented:
> Another option would be to go back to the public Result<> fields approach, so folks can do if api_struct.my_optional_hook.is_ok() {}, but then it's back to being strongly tied to the libloading API :/

I don't understand the issue with putting fields public now, in order to allow callers to check for errors ahead of time.

I believe @nagisa [comment](https://github.com/rust-lang/rust-bindgen/pull/1846#issuecomment-678784459) was for when the `TestLib` generated structure contained public fields of type `libloading::Symbol` but after a few iterations and @timfish proposition, this has now been changed to having the fields as function pointers and the conversion done in the constructor.
What would be the impact of breaking changes in `libloading` if the fields as they are now were made public? And similarly to `bindgen`? The methods do contain exactly the same types exposed as the function pointers in the generated structure fields. My point being that I would think a breaking change to `libloading` or `bindgen` would impact the current implementation excatly the same way if the fields are public or not.

goto-bus-stop commented:
I hadn't noticed that the field types were changed! If they are just function pointers it would be fine to expose them, I think.

for me, having access to the raw pointers inside an `Option<>` or `Result<>` would be the ideal API. I guess it would be up to bindgen maintainers to decide if optional symbols are somethign they want to support at all :sweat_smile:

hug-dev commented:
I would also think that the cost of the check of errors is minimal in comparison to the loading of the symbols, especially in optimised builds. I would say that paying that cost seems more safe than forcing to panic but if this is really a problem there could be a configurable option and users of this feature could take the responsability of filtering the symbols they are sure exist with the `allowlist`/`blocklist` functionality.

Also, `libloading::Library::get` returns a range of [errors](https://docs.rs/libloading/0.6.3/libloading/enum.Error.html) and maybe not all of them are directly linked with the fact that a symbol is missing.

timfish commented:
I've got some manually crafted libloading code and the problem I've seen with exposing the function pointers as pub fields is that you cannot call them directly like:
```rust
some_struct.some_fn_ptr(arg);
```
You have to:
```rust
let some_fn_ptr = some_struct.some_fn_ptr;
some_fn_ptr(arg);
```
Does anybody know why this is?

I would prefer most loading errors to bubble up during initialisation and if missing symbols to be exposed, then use `Option`.

hug-dev commented:
> You have to:
> ```
> let some_fn_ptr = some_struct.some_fn_ptr;
> some_fn_ptr(arg);
> ```
Would it be because more parenthesis are needed 😋? As in: `(some_struct.some_fn_ptr)(arg)`. I can see that @joechrisellis has them in the methods but not sure if they are needed there.

In any case, if making fields public is not a problem we could have both things:
* being able to directly access the fields
* having the methods to make it easier

I would prefer returning a `Result` as there would be loss of error information during the conversion `Err(libloading::Error)` -> `None`. Callers that want to disrecard this could call `.ok()` on the `Result`?

timfish commented:
> Would it be because more parenthesis are needed 😋?

Probably! 🤦‍♂️ 🤣

But yes, my point was that the methods make the interface a bit simpler to consume!

Michael-F-Bryan commented:
Besides ergonomics, exposing the bindings as methods is required for safety. If the fields were public, it'd be too easy for people to extract the underlying function pointer and accidentally use it after the library is unloaded from memory. That's one of the big reasons `libloading` uses its own `Symbol` wrapper. 

See https://github.com/nagisa/rust_libloading/issues/13#issuecomment-218839531 from the `libloading` repo and the linked rust-lang/rust#29625 for more.

hug-dev commented:
Agree that having `Symbol` seems safer. So one option would be to have only public fields of type `Result<Symbol<..>>` and the methods to access the functions easily and that also return a `Result`.

I think also that what proposed @goto-bus-stop might be a good idea:
> I don't know what a good solution would be though—adding a method to check for symbol existence might be too much auto-generated API surface for bindgen, and you'd have to deal with the possibility of loaded symbol names clashing with the bindgen-provided method names…

For each function `toto` there could be two associated methods on the `TestLib` struct:
* `fn can_call_toto(&self) -> Result<(), libloading::Error>`: returning basically `self.toto.map(|_| ())`
* `fn toto`: expected prototype but without returning a `Result`, directly return the contained value or panic if there was an error when calling `get`

Then the assumption for callers would be to do something like:
```rust
can_call_toto()?;
toto();
```
Or directly call `toto` if you are sure!

>  the possibility of loaded symbol names clashing with the bindgen-provided method names…

Maybe we could address that by having the implementation of the `can_call` methods inside their own namespace. I believe you can have `impl` blocks inside submodules. Could even be the `can_call` module, with the same name as the function:
```rust
can_call::toto()?;
toto();
```

Michael-F-Bryan commented:
> Agree that having Symbol seems safer. 

@hug-dev, keep in mind that [`libloading::Symbol`](https://docs.rs/libloading/0.6.3/libloading/struct.Symbol.html) holds a reference to the original library to prevent accidentally using it after the library is unloaded without runtime overhead. Storing a `Symbol` in your wrapper struct alongside the `libloading::Library` would create a self-referential type. 

That's why the wrapper's constructor (as currently formulated) would dereference a `Symbol` to get at the function pointer directly, and store that in the generated type. 

Safety is maintained by hiding function pointers behind a method call and making sure the wrapper retains ownership of the `Library`, making sure functions and pointers to static variables can't outlive the library they came from.

> So one option would be to have only public fields of type Result<Symbol<..>> and the methods to access the functions easily and that also return a Result.

You *can't* make the fields public. 

If they are public, it's possible to use the fact that `Symbol` implements `Deref<Target=fn(...)>` and that function pointers are `Copy` to use a function pointer after the library it points into is removed from memory.

> For each function toto there could be two associated methods on the TestLib struct:
>
> -   fn can_call_toto(&self) -> Result<(), libloading::Error>: returning basically self.toto.map(|_| ())
> -   fn toto: expected prototype but without returning a Result, directly return the contained value or panic if there was an error when calling get

This approach seems to be the most ergonomic. That way you can look-before-you-leap if you *know* a symbol will always be present, but also perform the check if you need to. I'm guessing the `unwrap()` call won't add enough overhead to be noticeable considering you are already doing dynamic dispatch with the function call.

To help with namespacing, what about adding a helper struct so you can write `my_bindings.can_call().toto()`? So the `can_call()` method returns a helper type that borrows the bindings and has methods with the equivalent of `can_call_toto()`. 

In the unlikely event that a library has a `can_call` symbol, you could keep prepending/appending underscores until there are no ambiguities.

joechrisellis commented:
@Michael-F-Bryan has made some excellent points above. I also like @hug-dev's idea of implementing a `can_call`/look-before-you-leap method. Doing this via a helper struct seems like the right approach.

Thinking about the (probably) more common case where _all_ of the symbols in the dynamic library are guaranteed to be defined, I also think it would be useful to have an option that allows us to fail early if a `libloading::Error` is raised inside the `new` method for our dynamic loading library type. This way, we can avoid panicking at runtime if something goes wrong early on when we're trying to load the library. I think that this should be the default option; I would guess that most people aren't working with modular/partially-defined libraries, and in that case, they can deal with an interface that is more ergonomical. I was thinking perhaps `--fail-early`/`fail_early`, set to true by default, and with some disclaimer that it's possible to get panics if you choose to turn this off? Would love to hear your thoughts on this. 😄

hug-dev commented:
Thanks for the explanation, your points make sense and give arguments in favor of having extra helper methods to check the `Result` only.

I like the idea of the helper struct containing all the `can_call_...` methods!

I like the idea of the `--fail-early` option which will also make code generation simpler: the `TestLib` struct can then directly contain the function pointers and it is no longer need to have the helper struct with the `can_call` methods. Or they should still be there but always return `Ok(())` so that the interface generated does not change with/without the option.

I would agree of adding this option but not sure of what should be the default. If the default is failing late, then a `--fail-early` option makes sense but if failing early is the default then probably a `--fail-late` option would be better 😃

joechrisellis commented:
@emilio hello! I think this is ready for review now -- I think we can add the `--fail-early` flag in a future PR if you think it's worthwhile. 😄

emilio commented:
(sorry for the lag here, I've been a bit too busy lately :/)

I'll try to get to this this week. Thanks for the patience!

DominicD commented:
Hi I am using this (great work btw! 👍  ) and it does work however I have 2 questions/problems. 
1. In the example in the beginning the function dynamic_loading() is called. It seems to me there was some change because in my case it does not exist and it works without calling it.
2. The bindings are created every time I build. I am not sure if this is related to this pull request or if it is a general bindgen thing? Is there anything I can do to change that?

joechrisellis commented:
@DominicD thanks for the feedback! To answer your questions:

1. Things have evolved slightly since this patch was first submitted. The `.dynamic_loading({true,false})` builder method has been superseded by `.dynamic_library_name("...")`. If you specify this option, it'll generate code for dynamic loading. If you don't, it'll run bindgen for your usual static bindings. Thanks for pointing this out though -- it is a good thing to point out for others!

2. I believe this is a general bindgen thing. FWIW I have heard of some people generating the bindings _once_ (for each platform they need) using bindgen from the CLI. This might be a good choice for you. It has the advantage of quicker build times, (a) bindgen won't need to be run and (b) you won't have to pull in bindgen and its dependencies for a clean build.

emilio commented on src/ir/context.rs:
```diff
@@ -2412,6 +2428,70 @@ If you encounter an error missing from this list, please file an issue or a PR!"
+            .filter(|&(_, item)| {
+                // If the user has not chosen to do dynamic loading, then we have nothing to
+                // do.
+                if !self.options().dynamic_loading {
```
Seems like this check should really go at the beginning of the function rather than being done for every item.

joechrisellis replied:
Quite right -- will fix.

emilio commented on src/ir/context.rs:
```diff
@@ -2412,6 +2428,70 @@ If you encounter an error missing from this list, please file an issue or a PR!"
+                // If there is a whitelist and this function is not on it, don't add it.
+                if !self.options().whitelisted_functions.is_empty() &&
+                    !self.options().whitelisted_functions.matches(&name)
+                {
```
I would expect this to be accounted for via the regular whitelisting mechanism, why is this needed?

joechrisellis replied:
Yep, I think this can be simplified. I think I had the wrong idea of the relationship between `items`, `codegen_items`, and `whitelisted_items`. I've changed this to iterate through `self.items()`, but filter out anything whose item ID is not inside `codegen_items`. It seems to do the job -- is that what you had in mind?

emilio commented on src/ir/context.rs:
```diff
@@ -2412,6 +2428,70 @@ If you encounter an error missing from this list, please file an issue or a PR!"
+                // If there is a blacklist and this function is on it, don't add it.
+                if !self.options().blacklisted_functions.is_empty() &&
+                    self.options.blacklisted_functions.matches(&name)
+                {
```
Same about this.

emilio commented on src/ir/context.rs:
```diff
@@ -2412,6 +2428,70 @@ If you encounter an error missing from this list, please file an issue or a PR!"
+
+                // We don't want to include the root module in the list of items to generate
+                // dynamic bindings for.
+                if item.id() == self.root_module {
```
We're only generating functions so this check is redundant, right?

emilio commented on src/lib.rs:
```diff
@@ -1431,6 +1437,16 @@ impl Builder {
+        mut self,
+        dynamic_library_name: T,
+    ) -> Self {
+        self.options.dynamic_loading = true;
```
Do you really need to keep track of `self.options.dynamic_loading`? It seems you really just need a library name, so the `dynamic_loading` checks could just be `self.options.dynamic_library_name.is_some()`.

joechrisellis replied:
Fixed this.

emilio commented on src/codegen/mod.rs:
```diff
@@ -3948,11 +3959,45 @@ pub(crate) fn codegen(
         );
 
+        // If the set of items to generate dynamic bindings for is nonempty...
+        if !context.dyngen_items().is_empty() {
```
So basically... This is just effectively generating all the functions right now, is that right?

The way I'd have approached this would be something like, maybe:

 * Add an extra field to `CodegenResult` (`dynamic_functions`? / whatever)
 * Handle this in `CodeGenerator for Function` (just something like `if ctx.options().dynamic_loading { /* generate appropriate things */ }`, and instead of pushing to the regular `CodegenResult` field, push it to the dynamic functions, which you'd then handle here.

Is there any reason such an approach wouldn't work? It'd handle a lot more edge cases than this (template functions, etc, which this patch doesn't handle), and should also be much simpler / not need extra `ItemSet`s.

emilio reviewed (changes requested):
Finally got some time to go through this. Really sorry again for the lag.

I think I might be missing something about why you chose this approach rather than something simpler hanging off the existing code generation machinery? Off-hand it seems this patch just takes a whole separate code generation step, while it could (I _think_) be part of the current machinery quite easily, unless I'm missing something.

emilio commented on src/codegen/mod.rs:
```diff
@@ -4356,6 +4403,35 @@ mod utils {
         args
     }
 
+    pub fn fnsig_argument_identifiers(
```
Could you document what this returns? The signature is not super-self-descriptive.

emilio commented on src/codegen/mod.rs:
```diff
@@ -3948,11 +3978,28 @@ pub(crate) fn codegen(
             &(),
         );
 
+        if context.options().dynamic_library_name.is_some() {
```
nit: `if let Some(ref lib_name) = context.options().dynamic_library_name {`.

Then use that instead of unwrapping around.

emilio commented on src/codegen/mod.rs:
```diff
@@ -3948,11 +3978,28 @@ pub(crate) fn codegen(
+                    "Check",
+                    context.options().dynamic_library_name.as_ref().unwrap(),
+                ]
+                .join(""),
```
just `format!("Check{}", lib_name)`?

emilio commented on src/codegen/dyngen.rs:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    pub fn add_function(
```
Maybe just `push` in order to match the result function name?

emilio commented on src/codegen/dyngen.rs:
```diff
@@ -0,0 +1,181 @@
+    init_fields: Vec<proc_macro2::TokenStream>,
+}
+
+impl Default for DynamicItems {
```
Derive this?

emilio reviewed (approved):
Sorry for the lag, didn't notice this was updated :(. Feel free to ping next time as needed, I get way too much github mail...

This looks basically good to me, thank you! I have some relatively minor nits, but I think with those addressed this looks good to me.

Given we need to do a breaking release anyhow for f2214790c8e6a5ce2e599564b8f5340d7abca92f, if this is updated soonish we can get it into 0.56.

Or alternatively, if you don't have the time, let me know and I can fix up the review comments myself.

The other great thing to do would be to have a documentation section in the book about this, how to use it and so on.

I don't think I want to block on that but I think it'd be great to add if any of you have the cycles.

Thanks again and sorry for the lag.

jsimmons commented on src/codegen/dyngen.rs:
```diff
@@ -0,0 +1,181 @@
+        ret: proc_macro2::TokenStream,
+        ret_ty: proc_macro2::TokenStream,
+    ) {
+        assert_eq!(args.len(), args_identifiers.len());
```
Just a note, this seems to assert on varargs parameters.

Not sure if those are not supported widely and this is expected? At least things *seem* to work when I remove the assert. The following snippet will trigger the assert.

```
int thing(int, ...);
```

emilio commented:
I pushed this with my review comments addressed. Will send a follow-up for the var-args issue mentioned by @jsimmons above.

emilio commented:
Thanks for all your work @joechrisellis (and everyone else who contributed and helped flesh out issues!)

hug-dev commented:
Nice!! 🥳 🥳 🥳 
Thanks @joechrisellis , @emilio and everybody involved!

emilio commented:
https://github.com/rust-lang/rust-bindgen/pull/1931 addresses the varargs functions (requires a bit of an API change but I think it's fine and the result is simpler).
